src/bytes.rs

```diff
@@ -200,6 +201,53 @@ impl Bytes {
         }
     }
 
+    /// Create [Bytes] with a buffer whose lifetime is controlled
+    /// via an explicit owner.
+    ///
+    /// A common use case is to zero-copy construct from mapped memory.
+    ///
+    /// ```ignore
+    /// use bytes::Bytes;
+    /// use std::fs::File;
+    /// use memmap2::Map;
+    ///
+    /// let file = File::open("upload_bundle.tar.gz")?;
+    /// let mmap = unsafe { Mmap::map(&file) }?;
+    /// let b = Bytes::from_owner(mmap);
+    /// ```
+    ///
+    /// The `owner` will be transferred to the constructed [Bytes] object, which
+    /// will ensure it is dropped once all remaining clones of the constructed
+    /// object are dropped. The owner will then be responsible for dropping the
+    /// specified region of memory as part of its [Drop] implementation.
+    ///
+    /// Note that converting [Bytes] constructed from an owner into a [BytesMut]
+    /// will always create a deep copy of the buffer into newly allocated memory.
+    pub fn from_owner<T: AsRef<[u8]>>(owner: T) -> Self {
```

Comment:
I wonder if we may want our own trait instead of `AsRef<[u8]>` so we can add more methods in the future.

Reply:
`<opinionated>`

A custom trait is how I had originally started implementing this, but gave up early on for a few reasons:

* Worse ergonomics for custom trait:
  * Many things already implement `AsRef<[u8]>`, so `Bytes::from_owner(mmap);` is easy and straight-forward.
  * Conversely, a custom trait would require `Bytes::from_custom(MyMmapWrapper::new(mmap))`.
* Safety:
  * `AsRef<[u8]>` can be called during construction and only once. It's a trick that makes the `from_owner` API safe.
  * Any additional APIs would - in practice - need to be called during the lifetime of bytes. As an example `trait CustomOwner { fn try_into_mut(self) -> Result<impl CustomOwnerMut, Self>; }` (unless I'm wrong) would make it into an `unsafe trait`.
  * What's missing from the current approach is zero-copy conversion to `BytesMut`.
    The implementor of a trait that that exposed this would need to be very careful here to respect the `Vec<u8>` semantics. Context: If `Bytes` is not unique, the existing code creates a newly allocated `Vec<u8>` copy of the referenced memory. Unless due care is taken - in the specific case of an `Mmap` - this would likely lead an implementor to erroneously assume that they should create an `MmapMut`. This would be wrong, especially if the mmapped section is backed by a file as the resulting behaviour would sometimes update the contents of a file and sometimes not, depending on `Bytes`'s internal refcount.
* Feature creep:
  * By all means, I think a custom trait approach might work in the future for a `fn from_custom <T: CustomOwner>` and as a good way of exposing more of the vtable functionality, but it is not where the current pain point in the API is and there are many hurdles to get past that as has been shown in previous efforts to tackle https://github.com/tokio-rs/bytes/issues/437.

The full-fat custom use case would mostly cover - as far as I can tell - mixing `Bytes` created and managed via custom memory allocators (QuestDB would certainly care!), but I suspect there's a more elegant way to handle those anyway.

`</opinionated>`

.. that said, if you guys prefer that route I'm happy to amend my PR in that direction.